### PR TITLE
Feat/add recipe

### PR DESCRIPTION
**Pull request by licetrosales**

### CODE REVIEW

File: backend/src/main/java/com/github/licetrosales/backend/controller/RecipeController.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    @GetMapping("/recipes")
+    List<Recipe> getAllRecipes() {
+        return recipeService.getAllRecipes();
+    }
+
+    @PostMapping("/recipes")
+    Recipe addRecipe(@RequestBody Recipe recipe) {
+
+        return recipeService.addRecipe(recipe);
+    }
```
Comment (Flooooooooooorian) [inline backend/src/main/java/com/github/licetrosales/backend/controller/RecipeController.java]: GET and ADD are two different features and should be developed in different branches and different PRs

File: backend/src/main/java/com/github/licetrosales/backend/controller/RecipeController.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import com.github.licetrosales.backend.repo.RecipeRepo;
+import com.github.licetrosales.backend.service.IdRecipeService;
+import com.github.licetrosales.backend.service.RecipeService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/users/userId")
```
Comment (Flooooooooooorian) [inline backend/src/main/java/com/github/licetrosales/backend/controller/RecipeController.java:13]: The urls looks a bit off 🤔 Do you want to show all recipes of one user or all recipes?

File: backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        ProductCategory productCategory
+        ) {
+    Ingredient(
+            String name,
+            String quantity,
+            Unit unit,
+            Boolean isInShoppingList
+    ) {
+        this(null, name, quantity, unit, false, null);
+    }
```
Comment (Flooooooooooorian) [inline backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java:24]: Do you really need this constructor? For new Ingredients i would create a "NewIngredient" Model

File: backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package com.github.licetrosales.backend.model;
+
+import lombok.With;
+import org.springframework.data.annotation.Id;
+
+import java.util.Collections;
+@With
+public record Ingredient(
+        @Id
```
Comment (Flooooooooooorian) [inline backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java:9]: Are you saving your ingredients in an own collection or just inside the recipes? Then you dont need the @Id annotation.

File: backend/src/main/java/com/github/licetrosales/backend/service/IdRecipeService.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.github.licetrosales.backend.service;
+
+import org.springframework.stereotype.Service;
+
+import java.util.UUID;
+@Service
+public class IdRecipeService {
+    private String customIdPrefix;
```
Comment (Flooooooooooorian) [inline backend/src/main/java/com/github/licetrosales/backend/service/IdRecipeService.java]: If not used please remove

File: backend/src/test/java/com/github/licetrosales/backend/controller/RecipeControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    @DirtiesContext
+    void getAllRecipes_shouldReturnListOfRecipes_whenGalleryisNotEmpty() throws Exception {
+        mockMvc.perform(get("/api/users/userId/recipes"))
+                .andExpect(status().isOk())
+                .andExpect(content().json("""
+                        []
+                        """));
+
+
+    }
```
Comment (Flooooooooooorian) [inline backend/src/test/java/com/github/licetrosales/backend/controller/RecipeControllerTest.java:53]: This test does not do what it says 🤔 
"whenGalleryisNotEmpty"

File: backend/src/test/java/com/github/licetrosales/backend/controller/RecipeControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                                             "menuCategory": "MAIN_COURSE",
+                                             "garnish": "Salat"
+                                           }
+                                          """
+                        ))
+                .andExpect(status().isOk())
+                .andReturn();
+        String content = response.getResponse().getContentAsString();
+        Recipe result = objectMapper.readValue(content, Recipe.class);
+        Recipe expected = new Recipe(result.id(), result.name(), result.mealType(), result.source(), result.image(), result.ingredients(), result.prepTime(), result.preparation(), result.portions(), result.favorite(), result.dishTypeCategory(), result.recipeCategory(), result.menuCategory(), result.garnish());
```
Comment (Flooooooooooorian) [inline backend/src/test/java/com/github/licetrosales/backend/controller/RecipeControllerTest.java:90]: Dont compare the result with itself, if there is an error in the result you wont notice it

File: frontend/src/components/CreateRecipeForm.tsx [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+        //ingredients: [],
+        prepTime: "",
+        preparation: "",
+        //dishTypeCategory: DishTypeCategory.VEGGIE,
+        portions: 1,
+        favorite: false,
+        //recipeCategory: RecipeCategory.LOW_CARB,
+        //menuCategory: MenuCatefory.MAIN_COURSE,
+        garnish: ""
+    }
```
Comment (Flooooooooooorian) [inline frontend/src/components/CreateRecipeForm.tsx:36]: Remove commented code

File: frontend/src/components/CreateRecipeForm.tsx [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+        setRecipeWithoutEnums(emptyRecipeFormWithoutEnums)
+        setMealType(MealType.LUNCH)
+        setDishTypeCategory(DishTypeCategory.VEGGIE)
+        setRecipeCategory(RecipeCategory.LOW_CARB)
+        setMenuCategory(MenuCategory.MAIN_COURSE)
+    }
+
+    useEffect(() => {
+
+    }, [recipeWithoutEnums])
```
Comment (Flooooooooooorian) [inline frontend/src/components/CreateRecipeForm.tsx]: I think you can remove the useEffect? 😄

File: frontend/src/components/MyMagicLunchApp.tsx [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    const recipeBaseUrl = "/api/users/userId/recipes"
+
+    function getRecipes() {
+        axios.get(recipeBaseUrl)
+            .then(recipeGalleryResponse => {
+                const newRecipeGallery: Recipe [] = recipeGalleryResponse.data;
+                setRecipes(newRecipeGallery);
+            })
+            .catch(errorMessageResponse => {
+                console.log("There is an error by GET request: " + errorMessageResponse)
```
Comment (Flooooooooooorian) [inline frontend/src/components/MyMagicLunchApp.tsx]: I would use at least an console.error for the error message, I would recommend to show the error to the user aswell (Show Error text on website or use something like toastify) we can have a look at that later

File: frontend/src/components/MyMagicLunchApp.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+            })
+    }
+
+    function addRecipe(newRecipeWithoutId: Recipe) {
+        axios.post(recipeBaseUrl, newRecipeWithoutId)
+            .then(newRecipeResponse => {
+                console.log("Neue Rezept: " + newRecipeResponse.data)
```
Comment (Flooooooooooorian) [inline frontend/src/components/MyMagicLunchApp.tsx]: Remove console.log

File: frontend/src/model/Ingredient.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
+export type Ingredient = {
+    id?: string,
+    name: string,
+    quantity: string,
+    unit?: Unit | string,
+    productCategory?: ProductCategory | string,
+    isInShoppingList?: boolean,
+}
```
Comment (Flooooooooooorian) [inline frontend/src/model/Ingredient.ts:8]: I not familiar with your models, but are you sure you want all these values as optional?

File: frontend/src/model/Recipe.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
+import {Ingredient} from "./Ingredient";
+
+
+export type Recipe = {
+    id?: string,
```
Comment (Flooooooooooorian) [inline frontend/src/model/Recipe.ts:5]: Why is the id optional?

**Review by Flooooooooooorian — APPROVED** [review]
Good job writing these tests 👍 Left you some comments.
Im not sure if you realy want to commit all the images and the example http request?

File: backend/src/main/java/com/github/licetrosales/backend/controller/RecipeController.java (hunk 2 above)
Comment (licetrosales) [reply backend/src/main/java/com/github/licetrosales/backend/controller/RecipeController.java:13]: The user have access only to his recipes. Here I want to show the user (given that the user is registered) all the recipes that are in his gallery.

File: backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java (hunk 3 above)
Comment (licetrosales) [reply backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java:24]: I plan to take the ingredient and add it as an item in the shopping list (that will be made from the recipe ingredients). I have to thing about that.

File: backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java (hunk 4 above)
Comment (licetrosales) [reply backend/src/main/java/com/github/licetrosales/backend/model/Ingredient.java:9]: I am planning to save the ingredients inside the recipe collection and copy them in a shopping list collection. The two collections (recipe & shopping list will be inside the user collection). Is the logic ok?
